### DISCUSSION TITLE
Testing the PullItemsAsync() method...

### DISCUSSION THREAD

**Discussion by blmiles** [Azure/azure-mobile-apps#769] [Q&A] [answered]
I am testing on a client device with a clean local db that get's created on app startup.

When I execute a table.PullItemsAsync() call, it won't Push as there is nothing to push. but it does not retrieve any records.
Expect 7 records which are in the azure db.
No issues cited in the Application Event Logs on the App Service.

If I add same 7 records (same Id's) from client-side and try a table.PushItemsAsync(), I get a series of errors in the Application Event Logs: _System.Data.SqlTypes.SqlNullValueException: Data is Null. This method or property cannot be called on Null values._

I assume (for now) this is because the "Version" field on the client-side is null (as they're newly created) and the server-side has a  version value for each record.

**Q. How to more effectively test a PullItemsAsync() call?  Possibly from outside the App...**

I'll deal with the push issues later...

ANYONE? hellooooo :)

Thank you!

**Comment by adrianhall** [comment]
The problem is really with the pull - I wouldn't expect the same 7 records to be added from the client-side and a push would work.

You will need to go down to the HTTP traces to figure out what is going on.  There isn't enough information here to diagnose.

**Reply by blmiles** [reply]
Oh I agree, I'm just struggling to dig into it properly to test and find the issue.

I already have common data in the cloud db that all clients get when they first start the app.
That happens through a pull into an empty db.
Because that first pull fails (returns nothing), the app inserts those records on the client-side with the same record IDs but on any subsequent sync, the push of those client-added records fails and probably rightly so because albeit the Id's are the same, the client-side version is null (being new) causing a null data failure.
Consequently, my whole sync execution code falls out of the try{} and into the catch{}, none of the pulls on any other data are hit.

I know this is terrible and UGLY, but I wonder if I should add same version number to the client-side records just to see if that is blocking the push?

Still, testing the service properly is a pain in the A :)

My live app is working. the new Maui app and service is what's failing. Same db.

**Reply by adrianhall** [reply, marked as answer]
You should not be pushing data into the database when the pull has failed. 

At the end, you really need to figure out WHY the pull is failing.  That means introducing logging on both sides and figuring it out.  The fact that the pull returns nothing but is successful indicates that the incremental sync is not working as expected.  Go back to first principals here - what's the HTTP request that is being sent?  Put that into Postman and make sure it returns the 7 records you expect.  If it doesn't, your problem is on the server side; if it does, then your problem is on the client side.  Now you get to step through the code and figure out WHY what is happening is actually happening.

**Reply by blmiles** [reply]
Yep, you're right. And pretty sure if I solve the first, it'll solve the 2nd issue too...

Thanks @adrianhall !!

**Comment by blmiles** [comment]
@adrianhall 

Following a bit of testing and messing with it all...some findings.

My live version of db has null values for UserID on standard common records all users can pull from db. It uses your older framework and it works too.

Since one cannot do this: ...CreateQuery().Where(ct => **ct.UserId==null** && ct.SportId == Settings.SportType);
I manually added an id into each record on the server side db to see if that would fix the Pull request..
**Voila! It pulls those tables.**

I'll need to manually add an ID to the common records in the NEW db before going live in Prod.

**Then there's the push...**

If I add a new record through the app and try sync db, I get a Server 500 error on the PushResults and the Application Event Logs in Azure App Service give me this (see attached file):

[TCAppErrorLogs.txt](https://github.com/Azure/azure-mobile-apps/files/12775909/TCAppErrorLogs.txt)
[TCAppErrorLogs2.txt](https://github.com/Azure/azure-mobile-apps/files/12777753/TCAppErrorLogs2.txt)

I've not knowingly added ANY triggers to my Db and no idea if any get added by EF or when deploying into Azure.
I did find this link and there's some code under the **Mitigation** heading...
https://learn.microsoft.com/en-us/ef/core/what-is-new/ef-core-7.0/breaking-changes?tabs=v7#sqlserver-tables-with-triggers

When you have any time, any advice would be appreciated.

Thanks!

**Comment by blmiles** [comment]
Hello Adrian,

Following a bit of testing and messing with it all...some findings.
I did post this in discussion for other's comment but unfortunately timing
is of the essence, Hence this email IF you could comment.

My live version of db has null values for UserID on standard common records
all users can pull from db. It uses your older framework and it works too.

Since one cannot do this: ...CreateQuery().Where(ct => *ct.UserId==null* &&
ct.SportId == Settings.SportType);
I manually added an id into each record on the server side db to see if
that would fix the Pull request..
*Voila! It pulls those tables.*

I'll need to manually add an ID to the common records in the NEW db before
going live in Prod.

*Then there's the push...*

If I add a new record through the app and try sync db, I get a Server 500
error on the PushResults and the Application Event Logs in Azure App
Service give me this (see attached file):

TCAppErrorLogs.txt
<https://github.com/Azure/azure-mobile-apps/files/12775909/TCAppErrorLogs.txt>
TCAppErrorLogs2.txt
<https://github.com/Azure/azure-mobile-apps/files/12777753/TCAppErrorLogs2.txt>

I've not knowingly added ANY triggers to my Db and no idea if any get added
by EF or when deploying into Azure.
I did find this link and there's some code under the *Mitigation* heading...
https://learn.microsoft.com/en-us/ef/core/what-is-new/ef-core-7.0/breaking-changes?tabs=v7#sqlserver-tables-with-triggers

When you have any time, any advice would be appreciated.

Thanks!

On Wed, Sep 27, 2023 at 12:48 PM Adrian Hall ***@***.***>
wrote:

> You should not be pushing data into the database when the pull has failed.
>
> At the end, you really need to figure out WHY the pull is failing. That
> means introducing logging on both sides and figuring it out. The fact that
> the pull returns nothing but is successful indicates that the incremental
> sync is not working as expected. Go back to first principals here - what's
> the HTTP request that is being sent? Put that into Postman and make sure it
> returns the 7 records you expect. If it doesn't, your problem is on the
> server side; if it does, then your problem is on the client side. Now you
> get to step through the code and figure out WHY what is happening is
> actually happening.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/Azure/azure-mobile-apps/discussions/769#discussioncomment-7125673>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AGIKTCBFRLPTLQMPLXVLUDTX4RKGBANCNFSM6AAAAAA5G7A3RM>
> .
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
>

Category: Microsoft.EntityFrameworkCore.Database.Command
EventId: 20102
SpanId: b091d8390c482e29
TraceId: cfd9ff7a951f79c639e2ee660c10e5ee
ParentId: 0000000000000000
RequestId: 80003145-0000-fd00-b63f-84710c7967bb
RequestPath: /tables/category
ActionId: 06f9012e-08cb-461f-89a8-fe08c5008cfb
ActionName: MyCompanyDataSyncServer.Controllers.CategoryController.CreateAsync (MyCompanyDataSyncServer)

Failed executing DbCommand (0ms) ***@***.***='?' (Size = 450), @p1='?' (DbType = Boolean), @p2='?' (Size = 4000), @p3='?' (Size = 4000), @p4='?' (Size = 4000), @p5='?' (DbType = DateTimeOffset), @p6='?' (Size = 4000)], CommandType='Text', CommandTimeout='30']
SET IMPLICIT_TRANSACTIONS OFF;
SET NOCOUNT ON;
INSERT INTO [Categories] ([Id], [Deleted], [Description], [Name], [SportId], [UpdatedAt], [UserId])
OUTPUT INSERTED.[Version]
VALUES ***@***.***, @p1, @p2, @p3, @p4, @p5, @p6);
Category: Microsoft.EntityFrameworkCore.Update
EventId: 10000
SpanId: 9adfba4a853650bb
TraceId: 02b1b46aeecb663edcb07eaf977e1e47
ParentId: 0000000000000000
RequestId: 8000400f-0000-f600-b63f-84710c7967bb
RequestPath: /tables/category
ActionId: 06f9012e-08cb-461f-89a8-fe08c5008cfb
ActionName: MyCompanyDataSyncServer.Controllers.CategoryController.CreateAsync (MyCompanyDataSyncServer)

An exception occurred in the database while saving changes for context type 'MyCompanyDataSyncServer.Db.AppDbContext'.
Microsoft.EntityFrameworkCore.DbUpdateException: Could not save changes because the target table has database triggers. Please configure your entity type accordingly, see https://aka.ms/efcore-docs-sqlserver-save-changes-and-triggers for more information.
 ---> Microsoft.Data.SqlClient.SqlException (0x80131904): The target table 'Categories' of the DML statement cannot have any enabled triggers if the statement contains an OUTPUT clause without INTO clause.
   at Microsoft.Data.SqlClient.SqlCommand.<>c.<ExecuteDbDataReaderAsync>b__209_0(Task`1 result)
   at System.Threading.Tasks.ContinuationResultTaskFromResultTask`2.InnerInvoke()
   at System.Threading.Tasks.Task.<>c.<.cctor>b__272_0(Object obj)
   at System.Threading.ExecutionContext.RunInternal(ExecutionContext executionContext, ContextCallback callback, Object state)
--- End of stack trace from previous location ---
   at System.Threading.ExecutionContext.RunInternal(ExecutionContext executionContext, ContextCallback callback, Object state)
   at System.Threading.Tasks.Task.ExecuteWithThreadLocal(Task& currentTaskSlot, Thread threadPoolThread)
--- End of stack trace from previous location ---
   at Microsoft.EntityFrameworkCore.Storage.RelationalCommand.ExecuteReaderAsync(RelationalCommandParameterObject parameterObject, CancellationToken cancellationToken)
   at Microsoft.EntityFrameworkCore.Storage.RelationalCommand.ExecuteReaderAsync(RelationalCommandParameterObject parameterObject, CancellationToken cancellationToken)
   at Microsoft.EntityFrameworkCore.Update.ReaderModificationCommandBatch.ExecuteAsync(IRelationalConnection connection, CancellationToken cancellationToken)
ClientConnectionId:ecffc866-9405-4bfa-837d-cf07442cebc3
Error Number:334,State:1,Class:16
ClientConnectionId before routing:592bcb2f-6e15-4d76-b7fd-bb4b7891f42c
Routing Destination:b8dbf3e5144f.tr34072.eastus1-a.worker.database.windows.net,11024
   --- End of inner exception stack trace ---
   at Microsoft.EntityFrameworkCore.SqlServer.Update.Internal.SqlServerModificationCommandBatch.ExecuteAsync(IRelationalConnection connection, CancellationToken cancellationToken)
   at Microsoft.EntityFrameworkCore.Update.Internal.BatchExecutor.ExecuteAsync(IEnumerable`1 commandBatches, IRelationalConnection connection, CancellationToken cancellationToken)
   at Microsoft.EntityFrameworkCore.Update.Internal.BatchExecutor.ExecuteAsync(IEnumerable`1 commandBatches, IRelationalConnection connection, CancellationToken cancellationToken)
   at Microsoft.EntityFrameworkCore.Update.Internal.BatchExecutor.ExecuteAsync(IEnumerable`1 commandBatches, IRelationalConnection connection, CancellationToken cancellationToken)
   at Microsoft.EntityFrameworkCore.ChangeTracking.Internal.StateManager.SaveChangesAsync(IList`1 entriesToSave, CancellationToken cancellationToken)
   at Microsoft.EntityFrameworkCore.ChangeTracking.Internal.StateManager.SaveChangesAsync(StateManager stateManager, Boolean acceptAllChangesOnSuccess, CancellationToken cancellationToken)
   at Microsoft.EntityFrameworkCore.SqlServer.Storage.Internal.SqlServerExecutionStrategy.ExecuteAsync[TState,TResult](TState state, Func`4 operation, Func`4 verifySucceeded, CancellationToken cancellationToken)
   at Microsoft.EntityFrameworkCore.DbContext.SaveChangesAsync(Boolean acceptAllChangesOnSuccess, CancellationToken cancellationToken)

**Comment by adrianhall** [comment]
So, null handling across the wire with text based encodings is a problem.  I’d recommend using empty strings instead.

That means you want to update your DB (as you have done), and write a custom repository or an access control provider that sets the UserId to the value you want on the server side prior to storage in the database.

Get Outlook for iOS<https://aka.ms/o0ukef>
________________________________

**Comment by blmiles** [comment]
Yep, kinda where I was going. Needed a concrete test first...

What can you tell me about the triggers and handling or negating them?
More info in rest of my last email...

Lindsay

**Comment by adrianhall** [comment]
We don't explicitly install any triggers since that would cause us to choose winners in databases.  However, we do include a [Timestamp]​ attribute in the EntityTableData class; this, in turn, installs a trigger to update the Version field on Create/Update.

The EntityTableRepository updates UpdatedAt (manually) when you write to the repository, before being stored in the database.  In addition, it sets the Id field to a Guid string automatically if it is set to null.  However, that is done in the repository code - not in the database.

So the triggers are really minimal.


Adrian Hall

Principal PM/Architect, Enterprise App Patterns
***@***.******@***.***>



________________________________

**Comment by blmiles** [comment]
Ok, any idea why I'm getting this error on the triggers then?
Two error files attached...

Failure to do with Trigger and then failed executing DbCommand. Or failed
command because of trigger...

Else I'll try follow this as I'm not sure how else to deal with it.
However, negating a trigger isn't a real fix either.
https://learn.microsoft.com/en-us/ef/core/what-is-new/ef-core-7.0/breaking-changes?tabs=v7#sqlserver-tables-with-triggers

Thx,

Lindsay





On Tue, Oct 3, 2023 at 2:18 PM Adrian Hall ***@***.***> wrote:

> We don't explicitly install any triggers since that would cause us to
> choose winners in databases. However, we do include a [Timestamp]​
> attribute in the EntityTableData class; this, in turn, installs a trigger
> to update the Version field on Create/Update.
>
> The EntityTableRepository updates UpdatedAt (manually) when you write to
> the repository, before being stored in the database. In addition, it sets
> the Id field to a Guid string automatically if it is set to null. However,
> that is done in the repository code - not in the database.
>
> So the triggers are really minimal.
>
>
> Adrian Hall
>
> Principal PM/Architect, Enterprise App Patterns
> ***@***.******@***.***>
>
>
>
> ________________________________
> From: blmiles ***@***.***>
> Sent: Tuesday, October 3, 2023 10:03 AM
> To: Azure/azure-mobile-apps ***@***.***>
> Cc: Adrian Hall ***@***.***>; Mention ***@***.***>
> Subject: Re: [Azure/azure-mobile-apps] Testing the PullItemsAsync()
> method... (Discussion #769)
>
>
> Yep, kinda where I was going. Needed a concrete test first...
>
> What can you tell me about the triggers and handling or negating them?
> More info in rest of my last email...
>
> Lindsay
>
> On Tue, Oct 3, 2023 at 1:01 PM Adrian Hall ***@***.***> wrote:
>
> > So, null handling across the wire with text based encodings is a
> problem.
> > I’d recommend using empty strings instead.
> >
> > That means you want to update your DB (as you have done), and write a
> > custom repository or an access control provider that sets the UserId to
> the
> > value you want on the server side prior to storage in the database.
> >
> > Get Outlook for iOS<https://aka.ms/o0ukef>
> > ________________________________
> > From: blmiles ***@***.***>
> > Sent: Tuesday, October 3, 2023 9:54:56 AM
> > To: Azure/azure-mobile-apps ***@***.***>
> > Cc: Adrian Hall ***@***.***>; Mention ***@***.***>
> > Subject: Re: [Azure/azure-mobile-apps] Testing the PullItemsAsync()
> > method... (Discussion #769)
> >
> >
> > Hello Adrian,
> >
> > Following a bit of testing and messing with it all...some findings.
> > I did post this in discussion for other's comment but unfortunately
> timing
> > is of the essence, Hence this email IF you could comment.
> >
> > My live version of db has null values for UserID on standard common
> > records
> > all users can pull from db. It uses your older framework and it works
> too.
> >
> > Since one cannot do this: ...CreateQuery().Where(ct => *ct.UserId==null*
> > &&
> > ct.SportId == Settings.SportType);
> > I manually added an id into each record on the server side db to see if
> > that would fix the Pull request..
> > *Voila! It pulls those tables.*
> >
> > I'll need to manually add an ID to the common records in the NEW db
> before
> > going live in Prod.
> >
> > *Then there's the push...*
> >
> > If I add a new record through the app and try sync db, I get a Server
> 500
> > error on the PushResults and the Application Event Logs in Azure App
> > Service give me this (see attached file):
> >
> > TCAppErrorLogs.txt
> > <
> >
> https://github.com/Azure/azure-mobile-apps/files/12775909/TCAppErrorLogs.txt>
>
> >
> > TCAppErrorLogs2.txt
> > <
> >
> https://github.com/Azure/azure-mobile-apps/files/12777753/TCAppErrorLogs2.txt>
>
> >
> >
> > I've not knowingly added ANY triggers to my Db and no idea if any get
> > added
> > by EF or when deploying into Azure.
> > I did find this link and there's some code under the *Mitigation*
> > heading...
> >
> >
> https://learn.microsoft.com/en-us/ef/core/what-is-new/ef-core-7.0/breaking-changes?tabs=v7#sqlserver-tables-with-triggers
> >
> > When you have any time, any advice would be appreciated.
> >
> > Thanks!
> >
> > On Wed, Sep 27, 2023 at 12:48 PM Adrian Hall ***@***.***>
> > wrote:
> >
> > > You should not be pushing data into the database when the pull has
> > failed.
> > >
> > > At the end, you really need to figure out WHY the pull is failing.
> That
> > > means introducing logging on both sides and figuring it out. The fact
> > that
> > > the pull returns nothing but is successful indicates that the
> > incremental
> > > sync is not working as expected. Go back to first principals here -
> > what's
> > > the HTTP request that is being sent? Put that into Postman and make
> sure
> > it
> > > returns the 7 records you expect. If it doesn't, your problem is on
> the
> > > server side; if it does, then your problem is on the client side. Now
> > you
> > > get to step through the code and figure out WHY what is happening is
> > > actually happening.
> > >
> > > —
> > > Reply to this email directly, view it on GitHub
> > > <
> >
> https://github.com/Azure/azure-mobile-apps/discussions/769#discussioncomment-7125673>,
>
> >
> > > or unsubscribe
> > > <
> >
> https://github.com/notifications/unsubscribe-auth/AGIKTCBFRLPTLQMPLXVLUDTX4RKGBANCNFSM6AAAAAA5G7A3RM>
>
> >
> > > .
> > > You are receiving this because you authored the thread.Message ID:
> > > ***@***.***>
> > >
> >
> > Category: Microsoft.EntityFrameworkCore.Database.Command
> > EventId: 20102
> > SpanId: b091d8390c482e29
> > TraceId: cfd9ff7a951f79c639e2ee660c10e5ee
> > ParentId: 0000000000000000
> > RequestId: 80003145-0000-fd00-b63f-84710c7967bb
> > RequestPath: /tables/category
> > ActionId: 06f9012e-08cb-461f-89a8-fe08c5008cfb
> > ActionName:
> > MyCompanyDataSyncServer.Controllers.CategoryController.CreateAsync
> > (MyCompanyDataSyncServer)
> >
> > Failed executing DbCommand (0ms) ***@***.***='?' (Size = 450), @p1='?'
> > (DbType = Boolean), @p2='?' (Size = 4000), @p3='?' (Size = 4000),
> @p4='?'
> > (Size = 4000), @p5='?' (DbType = DateTimeOffset), @p6='?' (Size =
> 4000)],
> > CommandType='Text', CommandTimeout='30']
> > SET IMPLICIT_TRANSACTIONS OFF;
> > SET NOCOUNT ON;
> > INSERT INTO [Categories] ([Id], [Deleted], [Description], [Name],
> > [SportId], [UpdatedAt], [UserId])
> > OUTPUT INSERTED.[Version]
> > VALUES ***@***.***, @p1, @p2, @p3, @p4, @p5, @p6);
> > Category: Microsoft.EntityFrameworkCore.Update
> > EventId: 10000
> > SpanId: 9adfba4a853650bb
> > TraceId: 02b1b46aeecb663edcb07eaf977e1e47
> > ParentId: 0000000000000000
> > RequestId: 8000400f-0000-f600-b63f-84710c7967bb
> > RequestPath: /tables/category
> > ActionId: 06f9012e-08cb-461f-89a8-fe08c5008cfb
> > ActionName:
> > MyCompanyDataSyncServer.Controllers.CategoryController.CreateAsync
> > (MyCompanyDataSyncServer)
> >
> > An exception occurred in the database while saving changes for context
> > type 'MyCompanyDataSyncServer.Db.AppDbContext'.
> > Microsoft.EntityFrameworkCore.DbUpdateException: Could not save changes
> > because the target table has database triggers. Please configure your
> > entity type accordingly, see
> > https://aka.ms/efcore-docs-sqlserver-save-changes-and-triggers for more
> > information.
> > ---> Microsoft.Data.SqlClient.SqlException (0x80131904): The target
> table
> > 'Categories' of the DML statement cannot have any enabled triggers if
> the
> > statement contains an OUTPUT clause without INTO clause.
> > at
> >
> Microsoft.Data.SqlClient.SqlCommand.<>c.<ExecuteDbDataReaderAsync>b__209_0(Task`1
>
> > result)
> > at
> >
> System.Threading.Tasks.ContinuationResultTaskFromResultTask`2.InnerInvoke()
> > at System.Threading.Tasks.Task.<>c.<.cctor>b__272_0(Object obj)
> > at System.Threading.ExecutionContext.RunInternal(ExecutionContext
> > executionContext, ContextCallback callback, Object state)
> > --- End of stack trace from previous location ---
> > at System.Threading.ExecutionContext.RunInternal(ExecutionContext
> > executionContext, ContextCallback callback, Object state)
> > at System.Threading.Tasks.Task.ExecuteWithThreadLocal(Task&
> > currentTaskSlot, Thread threadPoolThread)
> > --- End of stack trace from previous location ---
> > at
> >
> Microsoft.EntityFrameworkCore.Storage.RelationalCommand.ExecuteReaderAsync(RelationalCommandParameterObject
>
> > parameterObject, CancellationToken cancellationToken)
> > at
> >
> Microsoft.EntityFrameworkCore.Storage.RelationalCommand.ExecuteReaderAsync(RelationalCommandParameterObject
>
> > parameterObject, CancellationToken cancellationToken)
> > at
> >
> Microsoft.EntityFrameworkCore.Update.ReaderModificationCommandBatch.ExecuteAsync(IRelationalConnection
>
> > connection, CancellationToken cancellationToken)
> > ClientConnectionId:ecffc866-9405-4bfa-837d-cf07442cebc3
> > Error Number:334,State:1,Class:16
> > ClientConnectionId before routing:592bcb2f-6e15-4d76-b7fd-bb4b7891f42c
> > Routing Destination:
> > b8dbf3e5144f.tr34072.eastus1-a.worker.database.windows.net,11024
> > --- End of inner exception stack trace ---
> > at
> >
> Microsoft.EntityFrameworkCore.SqlServer.Update.Internal.SqlServerModificationCommandBatch.ExecuteAsync(IRelationalConnection
>
> > connection, CancellationToken cancellationToken)
> > at
> >
> Microsoft.EntityFrameworkCore.Update.Internal.BatchExecutor.ExecuteAsync(IEnumerable`1
>
> > commandBatches, IRelationalConnection connection, CancellationToken
> > cancellationToken)
> > at
> >
> Microsoft.EntityFrameworkCore.Update.Internal.BatchExecutor.ExecuteAsync(IEnumerable`1
>
> > commandBatches, IRelationalConnection connection, CancellationToken
> > cancellationToken)
> > at
> >
> Microsoft.EntityFrameworkCore.Update.Internal.BatchExecutor.ExecuteAsync(IEnumerable`1
>
> > commandBatches, IRelationalConnection connection, CancellationToken
> > cancellationToken)
> > at
> >
> Microsoft.EntityFrameworkCore.ChangeTracking.Internal.StateManager.SaveChangesAsync(IList`1
>
> > entriesToSave, CancellationToken cancellationToken)
> > at
> >
> Microsoft.EntityFrameworkCore.ChangeTracking.Internal.StateManager.SaveChangesAsync(StateManager
>
> > stateManager, Boolean acceptAllChangesOnSuccess, CancellationToken
> > cancellationToken)
> > at
> >
> Microsoft.EntityFrameworkCore.SqlServer.Storage.Internal.SqlServerExecutionStrategy.ExecuteAsync[TState,TResult](TState
>
> > state, Func`4 operation, Func`4 verifySucceeded, CancellationToken
> > cancellationToken)
> > at Microsoft.EntityFrameworkCore.DbContext.SaveChangesAsync(Boolean
> > acceptAllChangesOnSuccess, CancellationToken cancellationToken)
> >
> >
> > —
> > Reply to this email directly, view it on GitHub<
> >
> https://github.com/Azure/azure-mobile-apps/discussions/769#discussioncomment-7178135>,
>
> > or unsubscribe<
> >
> https://github.com/notifications/unsubscribe-auth/AALLUOJMLBFA4HR2XA2U53DX5Q7OBAVCNFSM6AAAAAA5G7A3ROVHI2DSMVQWIX3LMV43SRDJONRXK43TNFXW4Q3PNVWWK3TUHM3TCNZYGEZTK>.
>
> >
> > You are receiving this because you were mentioned.Message ID:
> ***@***.***>
> >
> > —
> > Reply to this email directly, view it on GitHub
> > <
> https://github.com/Azure/azure-mobile-apps/discussions/769#discussioncomment-7178200>,
>
> > or unsubscribe
> > <
> https://github.com/notifications/unsubscribe-auth/AGIKTCACAXCHAKBZDDKXNTLX5RAFFAVCNFSM6AAAAAA5G7A3ROVHI2DSMVQWIX3LMV43SRDJONRXK43TNFXW4Q3PNVWWK3TUHM3TCNZYGIYDA>
>
> > .
> > You are receiving this because you authored the thread.Message ID:
> > ***@***.***>
> >
>
> —
> Reply to this email directly, view it on GitHub<
> https://github.com/Azure/azure-mobile-apps/discussions/769#discussioncomment-7178225>,
> or unsubscribe<
> https://github.com/notifications/unsubscribe-auth/AALLUOPCL67ZOO4DNQRDBNDX5RAP3AVCNFSM6AAAAAA5G7A3ROVHI2DSMVQWIX3LMV43SRDJONRXK43TNFXW4Q3PNVWWK3TUHM3TCNZYGIZDK>.
>
> You are receiving this because you were mentioned.Message ID: ***@***.***>
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/Azure/azure-mobile-apps/discussions/769#discussioncomment-7178850>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AGIKTCHER65JQOD734RZQDDX5RJGZAVCNFSM6AAAAAA5G7A3ROVHI2DSMVQWIX3LMV43SRDJONRXK43TNFXW4Q3PNVWWK3TUHM3TCNZYHA2TA>
> .
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
>

Category: Microsoft.EntityFrameworkCore.Update
EventId: 10000
SpanId: 9adfba4a853650bb
TraceId: 02b1b46aeecb663edcb07eaf977e1e47
ParentId: 0000000000000000
RequestId: 8000400f-0000-f600-b63f-84710c7967bb
RequestPath: /tables/category
ActionId: 06f9012e-08cb-461f-89a8-fe08c5008cfb
ActionName: MyCompanyDataSyncServer.Controllers.CategoryController.CreateAsync (MyCompanyDataSyncServer)

An exception occurred in the database while saving changes for context type 'MyCompanyDataSyncServer.Db.AppDbContext'.
Microsoft.EntityFrameworkCore.DbUpdateException: Could not save changes because the target table has database triggers. Please configure your entity type accordingly, see https://aka.ms/efcore-docs-sqlserver-save-changes-and-triggers for more information.
 ---> Microsoft.Data.SqlClient.SqlException (0x80131904): The target table 'Categories' of the DML statement cannot have any enabled triggers if the statement contains an OUTPUT clause without INTO clause.
   at Microsoft.Data.SqlClient.SqlCommand.<>c.<ExecuteDbDataReaderAsync>b__209_0(Task`1 result)
   at System.Threading.Tasks.ContinuationResultTaskFromResultTask`2.InnerInvoke()
   at System.Threading.Tasks.Task.<>c.<.cctor>b__272_0(Object obj)
   at System.Threading.ExecutionContext.RunInternal(ExecutionContext executionContext, ContextCallback callback, Object state)
--- End of stack trace from previous location ---
   at System.Threading.ExecutionContext.RunInternal(ExecutionContext executionContext, ContextCallback callback, Object state)
   at System.Threading.Tasks.Task.ExecuteWithThreadLocal(Task& currentTaskSlot, Thread threadPoolThread)
--- End of stack trace from previous location ---
   at Microsoft.EntityFrameworkCore.Storage.RelationalCommand.ExecuteReaderAsync(RelationalCommandParameterObject parameterObject, CancellationToken cancellationToken)
   at Microsoft.EntityFrameworkCore.Storage.RelationalCommand.ExecuteReaderAsync(RelationalCommandParameterObject parameterObject, CancellationToken cancellationToken)
   at Microsoft.EntityFrameworkCore.Update.ReaderModificationCommandBatch.ExecuteAsync(IRelationalConnection connection, CancellationToken cancellationToken)
ClientConnectionId:ecffc866-9405-4bfa-837d-cf07442cebc3
Error Number:334,State:1,Class:16
ClientConnectionId before routing:592bcb2f-6e15-4d76-b7fd-bb4b7891f42c
Routing Destination:b8dbf3e5144f.tr34072.eastus1-a.worker.database.windows.net,11024
   --- End of inner exception stack trace ---
   at Microsoft.EntityFrameworkCore.SqlServer.Update.Internal.SqlServerModificationCommandBatch.ExecuteAsync(IRelationalConnection connection, CancellationToken cancellationToken)
   at Microsoft.EntityFrameworkCore.Update.Internal.BatchExecutor.ExecuteAsync(IEnumerable`1 commandBatches, IRelationalConnection connection, CancellationToken cancellationToken)
   at Microsoft.EntityFrameworkCore.Update.Internal.BatchExecutor.ExecuteAsync(IEnumerable`1 commandBatches, IRelationalConnection connection, CancellationToken cancellationToken)
   at Microsoft.EntityFrameworkCore.Update.Internal.BatchExecutor.ExecuteAsync(IEnumerable`1 commandBatches, IRelationalConnection connection, CancellationToken cancellationToken)
   at Microsoft.EntityFrameworkCore.ChangeTracking.Internal.StateManager.SaveChangesAsync(IList`1 entriesToSave, CancellationToken cancellationToken)
   at Microsoft.EntityFrameworkCore.ChangeTracking.Internal.StateManager.SaveChangesAsync(StateManager stateManager, Boolean acceptAllChangesOnSuccess, CancellationToken cancellationToken)
   at Microsoft.EntityFrameworkCore.SqlServer.Storage.Internal.SqlServerExecutionStrategy.ExecuteAsync[TState,TResult](TState state, Func`4 operation, Func`4 verifySucceeded, CancellationToken cancellationToken)
   at Microsoft.EntityFrameworkCore.DbContext.SaveChangesAsync(Boolean acceptAllChangesOnSuccess, CancellationToken cancellationToken)


Category: Microsoft.EntityFrameworkCore.Database.Command
EventId: 20102
SpanId: b091d8390c482e29
TraceId: cfd9ff7a951f79c639e2ee660c10e5ee
ParentId: 0000000000000000
RequestId: 80003145-0000-fd00-b63f-84710c7967bb
RequestPath: /tables/category
ActionId: 06f9012e-08cb-461f-89a8-fe08c5008cfb
ActionName: MyCompanyDataSyncServer.Controllers.CategoryController.CreateAsync (MyCompanyDataSyncServer)

Failed executing DbCommand (0ms) ***@***.***='?' (Size = 450), @p1='?' (DbType = Boolean), @p2='?' (Size = 4000), @p3='?' (Size = 4000), @p4='?' (Size = 4000), @p5='?' (DbType = DateTimeOffset), @p6='?' (Size = 4000)], CommandType='Text', CommandTimeout='30']
SET IMPLICIT_TRANSACTIONS OFF;
SET NOCOUNT ON;
INSERT INTO [Categories] ([Id], [Deleted], [Description], [Name], [SportId], [UpdatedAt], [UserId])
OUTPUT INSERTED.[Version]
VALUES ***@***.***, @p1, @p2, @p3, @p4, @p5, @p6);
